docs/modules/SDDs/pages/0021-cluster-catalog-compilation.adoc

```diff
@@ -0,0 +1,145 @@
+By using the Kapitan inventory the tool needs to figure out for which clusters the config changed and therefore needs a recompilation.
+In a first step this could be relatively rudimentary and only use the input information from a webhook (cluster/tenant ID) to decide which clusters to compile.
+
+The information if a cluster needs to be compiled is then stored in the cluster object of the Lieutenant API.
+In a first PoC phase this can be an annotation like `cluster.syn.tools/needs-compilation=true`.
+In a later stage this information should be part of the typed struct and might include further information like the exact version of the various input repos.
+
+
+==== Webhooks
+
+Triggering a webhook can be implemented in either a Git repo natively (for example https://developer.github.com/webhooks/[GitHub Webhooks]) or via a CI/CD system (for example a https://docs.gitlab.com/ee/ci/[GitLab CI] job which runs `curl`).
+A webhook can also be called by any other tool or even manually with a `curl` command.
+
+A webhook can include the following optional information:
+
+* Cluster ID for which to compile the catalog.
+This can be used in case where a specific cluster should be compiled
+
+* Tenant ID for which to figure out affected clusters.
+This can be used to limit the set of clusters to be considered
+
+* Set of classes that changed
+
+
+==== Authentication & Authorization
+
+To prevent exsessive resource usage or even DoS attacks, the webhooks need to be authenticated.
+A permission system needs to be in place to decide which entities are allowed to trigger which actions.
+It shouldn't be possible for example for a tenant to trigger compilation for clusters of other tenants.
+
+
+=== Compilation Jobs
+
+To compile the catalog for a cluster, a compilation job is created.
+The compilation job is a https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion[Kubernetes job] resource which executes the xref:0006-commodore-configuration-catalog-compilation.adoc[catalog compilation engine].
+
+To distribute the load this compilation generates and to follow a decentralized approach, these jobs are run on the same cluster as they compile the catalog for.
+The Steward component on a cluster starts a compilation job if the `cluster.syn.tools/needs-compilation` annotation is set to `true` on its own cluster object.
+It will then launch the respective job to compile the cluster catalog.
+
+For clusters where this isn't possible due to resource, security or other constraints, a central system runs the jobs.
+These clusters need to be configured in the central system so it will react on the mentioned annotation an create the compile jobs for them.
+
+
+=== User Stories
+
+==== Config Change in Global Config Repo
+
+If configuration in the global inventory repo changes, Quartermaster needs to find all clusters which are affected by the changed configuration.
```

Comment:
Probably a good idea to discuss the problems we're facing here, and sketching a first approach for actually determining which clusters are affected.

Having thought about this a bit more, I think we can get away with completely ignoring component classes, and just backtrace whether changes to a particular class affect a given cluster by constructing the cluster target and mocking all component classes as empty classes, and then listing all included classes using reclass.